File: examples/playground/ilovereact/native/src/hostConfig.ts
/**
 * React-reconciler host config for the native (QuickJS FFI) renderer.
 *
 * Ported from reactor01's renderer.js to TypeScript.
 * Emits mutation commands (CREATE, APPEND, UPDATE, etc.) that get serialized
 * as JSON and sent to Lua via __hostFlush.
 */

import type { HostConfig } from 'react-reconciler';
import { reportError } from './errorReporter';

// ── Types ────────────────────────────────────────────────

export interface Instance {
  id: number;
  type: string;
  props: Record<string, any>;
  handlers: Record<string, Function>;
  children: Instance[];
}

export interface TextInstance {
  id: number;
  text: string;
}

type Container = { id: number };
type Props = Record<string, any>;
type UpdatePayload = Record<string, any> | null;

interface Command {
  op: string;
  [key: string]: any;
}

// ── Globals / FFI bridge ─────────────────────────────────

declare const globalThis: {
  __hostFlush: (commands: string | Command[]) => void;
  __hostGetEvents: () => any[];
  _pollAndDispatchEvents?: () => void;
  [key: string]: any;
};

// ── Transport abstraction ────────────────────────────────

/** Injected flush handler — decouples the reconciler from a specific transport. */
let transportFlush: ((commands: Command[]) => void) | null = null;

/**
 * Register the transport flush handler.
 * Called by the bridge implementation (NativeBridge, WebSocketBridge, etc.)
 * to tell the reconciler how to deliver commands.
 */
export function setTransportFlush(fn: (commands: Command[]) => void): void {
  transportFlush = fn;
}

// ── State ────────────────────────────────────────────────

let nodeIdCounter = 0;
const pendingCommands: Command[] = [];
const rootInstances: Instance[] = [];

/**
 * Get the current root-level instances.
 * Used by remote targets (e.g., CC) to walk the JS-side tree
 * instead of sending mutation commands to a Lua bridge.
 */
export function getRootInstances(): Instance[] {
  return rootInstances;
}

/**
 * Registry of event handlers keyed by nodeId.
 * Handlers never cross the bridge — they stay in JS.
 * Lua sends events referencing targetId, and we dispatch here.
 */
export const handlerRegistry = new Map<number, Record<string, Function>>();

// ── Helpers ──────────────────────────────────────────────

function emit(cmd: Command): void {
  pendingCommands.push(cmd);
}

/**
 * Merge multiple UPDATE commands targeting the same node into a single command.
 * Non-UPDATE commands and UPDATEs for distinct nodes pass through unchanged.
 */
function coalesceCommands(commands: Command[]): Command[] {
  const updateMap = new Map<number, number>(); // nodeId -> index in output
  const output: Command[] = [];

  for (const cmd of commands) {
    if (cmd.op === 'UPDATE' && cmd.id != null) {
      const existingIdx = updateMap.get(cmd.id);
      if (existingIdx !== undefined) {
        const existing = output[existingIdx];
        // Merge props
        existing.props = { ...existing.props, ...cmd.props };
        // Merge style sub-objects
        if (existing.props.style && cmd.props.style) {
          existing.props.style = { ...existing.props.style, ...cmd.props.style };
        }
        // Merge removal arrays
        if (cmd.removeKeys) {
          existing.removeKeys = [...(existing.removeKeys || []), ...cmd.removeKeys];
        }
        if (cmd.removeStyleKeys) {
          existing.removeStyleKeys = [...(existing.removeStyleKeys || []), ...cmd.removeStyleKeys];
        }
        // Latest hasHandlers wins
        if (cmd.hasHandlers !== undefined) {
          existing.hasHandlers = cmd.hasHandlers;
        }
        continue;
      }
      updateMap.set(cmd.id, output.length);
    }
    output.push({ ...cmd });
  }

  return output;
}

export function flushToHost(): void {
  if (pendingCommands.length === 0) return;

  // Resolve transport on first flush (backwards compat: fall back to global)
  if (!transportFlush) {
    if (typeof globalThis.__hostFlush === 'function') {
      transportFlush = (cmds) => globalThis.__hostFlush(cmds);
    } else {
      return;
    }
  }

  const coalesced = coalesceCommands(pendingCommands);
  try {
    // Send as JSON string to avoid QuickJS GC race during FFI object traversal.
    // Large strings (500+ chars) can be silently collected by GC during property
    // enumeration of sibling properties, causing silent data loss across the bridge.
    transportFlush(JSON.stringify(coalesced));
  } catch (e) {
    reportError(e, 'flushToHost (' + coalesced.length + ' commands)');
  }
  pendingCommands.length = 0;
}

/**
 * Separate on* handler props from regular props.
 * Handlers stay in JS; only clean props cross the bridge.
 */
export function extractHandlers(
  props: Props
): { clean: Record<string, any>; handlers: Record<string, Function> } {
  const clean: Record<string, any> = {};
  const handlers: Record<string, Function> = {};

  for (const key of Object.keys(props)) {
    if (key === 'children') continue;
    if (key.startsWith('on') && typeof props[key] === 'function') {
      handlers[key] = props[key];
    } else {
      clean[key] = props[key];
    }
  }

  return { clean, handlers };
}

/**
 * Shallow equality check for prop diffing.
 */
export function shallowEqual(
  a: Record<string, any>,
  b: Record<string, any>
): boolean {
  const keysA = Object.keys(a);
  const keysB = Object.keys(b);
  if (keysA.length !== keysB.length) return false;
  for (const key of keysA) {
    if (a[key] !== b[key]) return false;
  }
  return true;
}

/**
 * Compare two handler maps by reference equality of each handler.
 */
function handlersEqual(a: Record<string, Function>, b: Record<string, Function>): boolean {
  const keysA = Object.keys(a);
  const keysB = Object.keys(b);
  if (keysA.length !== keysB.length) return false;
  for (const k of keysA) {
    if (a[k] !== b[k]) return false;
  }
  return true;
}

/**
 * Compute a partial diff between two style objects.
 * Returns changed keys (with new values) and removed keys.
 */
function diffStyleObjects(
  oldStyle: Record<string, any>,
  newStyle: Record<string, any>
): { changed: Record<string, any> | null; removed: string[] } {
  const changed: Record<string, any> = {};
  const removed: string[] = [];
  let hasChanged = false;

  for (const key of Object.keys(newStyle)) {
    if (oldStyle[key] !== newStyle[key]) {
      changed[key] = newStyle[key];
      hasChanged = true;
    }
  }

  for (const key of Object.keys(oldStyle)) {
    if (!(key in newStyle)) {
      removed.push(key);
    }
  }

  return {
    changed: hasChanged ? changed : null,
    removed,
  };
}

/**
 * Compute a minimal diff between two clean prop objects.
 * Returns the changed props (with partial style diff), removed top-level keys,
 * and removed style keys. Returns null if nothing changed.
 */
function diffCleanProps(
  oldClean: Record<string, any>,
  newClean: Record<string, any>
): { diff: Record<string, any>; removeKeys: string[]; removeStyleKeys: string[] } | null {
  const diff: Record<string, any> = {};
  const removeKeys: string[] = [];
  let removeStyleKeys: string[] = [];
  let hasDiff = false;

  for (const key of Object.keys(newClean)) {
    const oldVal = oldClean[key];
    const newVal = newClean[key];

    if (key === 'style') {
      const styleDiff = diffStyleObjects(oldVal || {}, newVal || {});
      if (styleDiff.changed) {
        diff.style = styleDiff.changed;
        hasDiff = true;
      }
      removeStyleKeys = styleDiff.removed;
      if (removeStyleKeys.length > 0) hasDiff = true;
    } else if (oldVal !== newVal) {
      diff[key] = newVal;
      hasDiff = true;
    }
  }

  for (const key of Object.keys(oldClean)) {
    if (key === 'style') continue; // handled above
    if (!(key in newClean)) {
      removeKeys.push(key);
      hasDiff = true;
    }
  }

  return hasDiff ? { diff, removeKeys, removeStyleKeys } : null;
}

// ── DefaultEventPriority ────────────────────────────────

const DefaultEventPriority = 0b0000000000000000000000000010000;

// ── Host Config ──────────────────────────────────────────

export const hostConfig: HostConfig<
  string,         // Type
  Props,          // Props
  Container,      // Container
  Instance,       // Instance
  TextInstance,    // TextInstance
  never,          // SuspenseInstance
  never,          // HydratableInstance
  Instance,       // PublicInstance
  {},             // HostContext
  UpdatePayload,  // UpdatePayload
  unknown,        // ChildSet
  number,         // TimeoutHandle
  number           // NoTimeout
> = {
  // ── Feature flags ────────────────────────────────────

  supportsMutation: true,
  supportsPersistence: false,
  supportsHydration: false,
  isPrimaryRenderer: true,

  // ── Context ──────────────────────────────────────────

  getRootHostContext() {
    return {};
  },

  getChildHostContext(_parentContext: {}) {
    return {};
  },

  // ── Instance creation ────────────────────────────────

  createInstance(
    type: string,
    props: Props,
    _rootContainer: Container,
    _hostContext: {},
    internalHandle?: any // React fiber (opaque, but we bend the rules for debugging)
  ): Instance {
    const id = ++nodeIdCounter;
    const { clean, handlers } = extractHandlers(props);

    const hasHandlers = Object.keys(handlers).length > 0;

    // Extract component debug info from fiber (dev tooling only)
    let debugName: string | undefined;
    let debugSource: { fileName?: string; lineNumber?: number } | undefined;

    if (internalHandle) {
      try {
        // Walk up the fiber tree to find the nearest component name
        let fiber = internalHandle;
        while (fiber) {
          // Function/class components have a type with a name or displayName
          if (fiber.type && typeof fiber.type === 'function') {
            debugName = fiber.type.displayName || fiber.type.name;
            break;
          }
          fiber = fiber.return;
        }

        // Try to get source location (requires JSX dev transform)
        if (internalHandle._debugSource) {
          debugSource = {
            fileName: internalHandle._debugSource.fileName,
            lineNumber: internalHandle._debugSource.lineNumber,
          };
        }
      } catch (e) {
        // Silently fail — fiber internals may change between React versions
      }
    }

    emit({
      op: 'CREATE',
      id,
      type,
      props: clean,
      hasHandlers,
      debugName,
      debugSource,
    });

    if (hasHandlers) {
      handlerRegistry.set(id, handlers);
    }

    return { id, type, props: clean, handlers, children: [] };
  },

  createTextInstance(text: string): TextInstance {
    const id = ++nodeIdCounter;
    emit({ op: 'CREATE_TEXT', id, text });
    return { id, text };
  },

  // ── Tree building (pre-commit) ──────────────────────

  appendInitialChild(parent: Instance, child: Instance | TextInstance) {
    (parent.children as any[]).push(child);
    emit({ op: 'APPEND', parentId: parent.id, childId: child.id });
  },

  finalizeInitialChildren() {
    return false;
  },

  // ── Mutations ────────────────────────────────────────

  appendChild(parent: Instance, child: Instance | TextInstance) {
    (parent.children as any[]).push(child);
    emit({ op: 'APPEND', parentId: parent.id, childId: child.id });
  },

  appendChildToContainer(container: Container, child: Instance | TextInstance) {
    rootInstances.push(child as Instance);
    emit({ op: 'APPEND_TO_ROOT', childId: child.id });
  },

  removeChild(parent: Instance, child: Instance | TextInstance) {
    const idx = (parent.children as any[]).indexOf(child);
    if (idx !== -1) (parent.children as any[]).splice(idx, 1);
    emit({ op: 'REMOVE', parentId: parent.id, childId: child.id });
    cleanupHandlers(child);
  },

  removeChildFromContainer(_container: Container, child: Instance | TextInstance) {
    const idx = rootInstances.indexOf(child as Instance);
    if (idx !== -1) rootInstances.splice(idx, 1);
    emit({ op: 'REMOVE_FROM_ROOT', childId: child.id });
    cleanupHandlers(child);
  },

  insertBefore(
    parent: Instance,
    child: Instance | TextInstance,
    before: Instance | TextInstance
  ) {
    const arr = parent.children as any[];
    const idx = arr.indexOf(before);
    if (idx !== -1) {
      arr.splice(idx, 0, child);
    } else {
      arr.push(child);
    }
    emit({
      op: 'INSERT_BEFORE',
      parentId: parent.id,
      childId: child.id,
      beforeId: before.id,
    });
  },

  insertInContainerBefore(
    _container: Container,
    child: Instance | TextInstance,
    before: Instance | TextInstance
  ) {
    const idx = rootInstances.indexOf(before as Instance);
    if (idx !== -1) {
      rootInstances.splice(idx, 0, child as Instance);
    } else {
      rootInstances.push(child as Instance);
    }
    emit({
      op: 'INSERT_BEFORE_ROOT',
      childId: child.id,
      beforeId: before.id,
    });
  },

  // ── Updates ──────────────────────────────────────────

  prepareUpdate(
    _instance: Instance,
    _type: string,
    oldProps: Props,
    newProps: Props
  ): UpdatePayload {
    const { clean: oldClean, handlers: oldH } = extractHandlers(oldProps);
    const { clean: newClean, handlers: newH } = extractHandlers(newProps);

    const handlersChanged = !handlersEqual(oldH, newH);
    const propsDiff = diffCleanProps(oldClean, newClean);

    if (!propsDiff && !handlersChanged) return null;

    if (!propsDiff && handlersChanged) {
      return { __handlersOnly: true };
    }

    return propsDiff; // { diff, removeKeys, removeStyleKeys }
  },

  commitUpdate(
    instance: Instance,
    updatePayload: UpdatePayload,
    _type: string,
    _oldProps: Props,
    newProps: Props
  ) {
    const { clean, handlers } = extractHandlers(newProps);

    // Update handler registry
    if (Object.keys(handlers).length > 0) {
      handlerRegistry.set(instance.id, handlers);
    } else {
      handlerRegistry.delete(instance.id);
    }

    instance.handlers = handlers;
    instance.props = clean;

    if (updatePayload && !(updatePayload as any).__handlersOnly) {
      const hasHandlers = Object.keys(handlers).length > 0;
      const payload = updatePayload as { diff: Record<string, any>; removeKeys: string[]; removeStyleKeys: string[] };

      const cmd: any = {
        op: 'UPDATE',
        id: instance.id,
        props: payload.diff,
        hasHandlers,
      };

      if (payload.removeKeys.length > 0) {
        cmd.removeKeys = payload.removeKeys;
      }
      if (payload.removeStyleKeys.length > 0) {
        cmd.removeStyleKeys = payload.removeStyleKeys;
      }

      emit(cmd);
    } else if (updatePayload && (updatePayload as any).__handlersOnly) {
      // Still need to inform Lua about hasHandlers change
      const hasHandlers = Object.keys(handlers).length > 0;
      emit({
        op: 'UPDATE',
        id: instance.id,
        props: {},
        hasHandlers,
      });
    }
  },

  commitTextUpdate(_textInstance: TextInstance, _oldText: string, newText: string) {
    _textInstance.text = newText;
    emit({ op: 'UPDATE_TEXT', id: _textInstance.id, text: newText });
  },

  // ── Commit lifecycle ─────────────────────────────────

  prepareForCommit() {
    return null;
  },

  resetAfterCommit() {
    flushToHost();
  },

  // ── Misc required methods ────────────────────────────

  shouldSetTextContent() {
    return false;
  },

  getPublicInstance(instance: Instance) {
    return instance;
  },

  preparePortalMount() {},

  scheduleTimeout: setTimeout,
  cancelTimeout: clearTimeout,
  noTimeout: -1,

  getCurrentEventPriority() {
    return DefaultEventPriority;
  },

  getInstanceFromNode() {
    return null;
  },

  beforeActiveInstanceBlur() {},
  afterActiveInstanceBlur() {},

  prepareScopeUpdate() {},
  getInstanceFromScope() {
    return null;
  },

  detachDeletedInstance() {},

  clearContainer() {},
};

// ── Internal helpers ─────────────────────────────────────

function cleanupHandlers(node: Instance | TextInstance): void {
  if ('id' in node) {
    handlerRegistry.delete(node.id);
  }
  if ('children' in node) {
    for (const child of (node as Instance).children) {
      cleanupHandlers(child);
    }
  }
}

File: renderer/hostConfig.lua
local M = {}

local function deep_clone(value, seen)
  if type(value) ~= "table" then
    return value
  end
  seen = seen or {}
  if seen[value] then
    return seen[value]
  end
  local out = {}
  seen[value] = out
  for k, v in pairs(value) do
    out[deep_clone(k, seen)] = deep_clone(v, seen)
  end
  return out
end

local function array_index_of(list, item)
  for i, value in ipairs(list or {}) do
    if value == item then
      return i
    end
  end
  return nil
end

local function remove_value(list, item)
  local idx = array_index_of(list, item)
  if idx then
    table.remove(list, idx)
  end
end

local function insert_before(list, item, before)
  remove_value(list, item)
  if before ~= nil then
    local idx = array_index_of(list, before)
    if idx then
      table.insert(list, idx, item)
      return
    end
  end
  table.insert(list, item)
end

local function sorted_keys(map)
  local keys = {}
  for key in pairs(map or {}) do
    keys[#keys + 1] = key
  end
  table.sort(keys)
  return keys
end

local function handler_signature(fn)
  local info = debug and debug.getinfo and debug.getinfo(fn, "Sln")
  if not info then
    return "(native)"
  end
  local source = info.short_src or info.source or "(unknown)"
  local line = info.linedefined or 0
  local last = info.lastlinedefined or line
  if last ~= line then
    return string.format("%s:%d-%d", source, line, last)
  end
  return string.format("%s:%d", source, line)
end

function M.extractHandlers(props)
  local clean = {}
  local handlers = {}

  for key, value in pairs(props or {}) do
    if key ~= "children" and type(key) == "string" and key:sub(1, 2) == "on" and type(value) == "function" then
      handlers[key] = value
    elseif key ~= "children" then
      clean[key] = value
    end
  end

  return clean, handlers
end

local function diff_style_objects(old_style, new_style)
  local changed = {}
  local removed = {}
  local has_changed = false

  old_style = old_style or {}
  new_style = new_style or {}

  for key, value in pairs(new_style) do
    if old_style[key] ~= value then
      changed[key] = value
      has_changed = true
    end
  end

  for key in pairs(old_style) do
    if new_style[key] == nil then
      removed[#removed + 1] = key
      has_changed = true
    end
  end

  if not has_changed then
    return nil, removed
  end

  return changed, removed
end

function M.diffCleanProps(old_clean, new_clean)
  local diff = {}
  local remove_keys = {}
  local remove_style_keys = {}
  local has_diff = false

  old_clean = old_clean or {}
  new_clean = new_clean or {}

  for key, new_value in pairs(new_clean) do
    local old_value = old_clean[key]
    if key == "style" then
      local style_diff, removed = diff_style_objects(old_value, new_value)
      if style_diff then
        diff.style = style_diff
        has_diff = true
      end
      remove_style_keys = removed
      if #remove_style_keys > 0 then
        has_diff = true
      end
    elseif old_value ~= new_value then
      diff[key] = new_value
      has_diff = true
    end
  end

  for key in pairs(old_clean) do
    if key ~= "style" and new_clean[key] == nil then
      remove_keys[#remove_keys + 1] = key
      has_diff = true
    end
  end

  if not has_diff then
    return nil
  end

  return {
    diff = diff,
    removeKeys = remove_keys,
    removeStyleKeys = remove_style_keys,
  }
end

function M.coalesceCommands(commands)
  local update_index = {}
  local output = {}

  for _, cmd in ipairs(commands or {}) do
    if cmd.op == "UPDATE" and cmd.id ~= nil then
      local existing_idx = update_index[cmd.id]
      if existing_idx ~= nil then
        local existing = output[existing_idx]
        local prev_style = existing.props and existing.props.style

        existing.props = existing.props or {}
        for key, value in pairs(cmd.props or {}) do
          existing.props[key] = deep_clone(value)
        end

        if prev_style and cmd.props and cmd.props.style then
          local merged = deep_clone(prev_style)
          for key, value in pairs(cmd.props.style) do
            merged[key] = deep_clone(value)
          end
          existing.props.style = merged
        end

        if cmd.removeKeys then
          existing.removeKeys = existing.removeKeys or {}
          for _, key in ipairs(cmd.removeKeys) do
            existing.removeKeys[#existing.removeKeys + 1] = key
          end
        end

        if cmd.removeStyleKeys then
          existing.removeStyleKeys = existing.removeStyleKeys or {}
          for _, key in ipairs(cmd.removeStyleKeys) do
            existing.removeStyleKeys[#existing.removeStyleKeys + 1] = key
          end
        end

        if cmd.hasHandlers ~= nil then
          existing.hasHandlers = cmd.hasHandlers
        end
        if cmd.handlerNames ~= nil then
          existing.handlerNames = deep_clone(cmd.handlerNames)
        end
      else
        update_index[cmd.id] = #output + 1
        output[#output + 1] = deep_clone(cmd)
      end
    else
      output[#output + 1] = deep_clone(cmd)
    end
  end

  return output
end

function M.buildHandlerMeta(handlers)
  local keys = sorted_keys(handlers)
  if #keys == 0 then
    return nil
  end

  local meta = {}
  for _, name in ipairs(keys) do
    meta[name] = handler_signature(handlers[name])
  end

  return meta
end

function M.newEmitter()
  local emitter = {
    pendingCommands = {},
    handlerRegistry = {},
    nodeIdCounter = 0,
  }

  function emitter:nextId()
    self.nodeIdCounter = self.nodeIdCounter + 1
    return self.nodeIdCounter
  end

  function emitter:emit(cmd)
    self.pendingCommands[#self.pendingCommands + 1] = cmd
  end

  function emitter:createText(text)
    local id = self:nextId()
    self:emit({ op = "CREATE_TEXT", id = id, text = tostring(text) })
    return id
  end

  function emitter:createInstance(type_name, props)
    local id = self:nextId()
    local clean, handlers = M.extractHandlers(props or {})
    local handler_names = sorted_keys(handlers)
    local cmd = {
      op = "CREATE",
      id = id,
      type = type_name,
      props = clean,
    }

    if #handler_names > 0 then
      cmd.hasHandlers = true
      cmd.handlerNames = handler_names
      cmd.handlerMeta = M.buildHandlerMeta(handlers)
    end

    self.handlerRegistry[id] = handlers
    self:emit(cmd)
    return id, clean, handlers
  end

  function emitter:append(parent_id, child_id)
    self:emit({ op = "APPEND", parentId = parent_id, childId = child_id })
  end

  function emitter:appendToRoot(child_id)
    self:emit({ op = "APPEND_TO_ROOT", childId = child_id })
  end

  function emitter:remove(parent_id, child_id)
    self:emit({ op = "REMOVE", parentId = parent_id, childId = child_id })
  end

  function emitter:removeFromRoot(child_id)
    self:emit({ op = "REMOVE_FROM_ROOT", childId = child_id })
  end

  function emitter:update(node_id, old_clean, new_clean, old_handlers, new_handlers)
    local diff = M.diffCleanProps(old_clean, new_clean)
    local old_names = sorted_keys(old_handlers)
    local new_names = sorted_keys(new_handlers)
    if not diff and #old_names == #new_names then
      local same = true
      for i = 1, #old_names do
        if old_names[i] ~= new_names[i] then
          same = false
          break
        end
      end
      if same then
        return nil
      end
    end

    local cmd = {
      op = "UPDATE",
      id = node_id,
      props = (diff and diff.diff) or {},
    }

    if diff then
      if #diff.removeKeys > 0 then
        cmd.removeKeys = diff.removeKeys
      end
      if #diff.removeStyleKeys > 0 then
        cmd.removeStyleKeys = diff.removeStyleKeys
      end
    end

    if #new_names > 0 then
      cmd.hasHandlers = true
      cmd.handlerNames = new_names
      cmd.handlerMeta = M.buildHandlerMeta(new_handlers)
    elseif #old_names > 0 then
      cmd.hasHandlers = false
      cmd.handlerNames = {}
      cmd.handlerMeta = nil
    end

    self.handlerRegistry[node_id] = new_handlers or {}
    self:emit(cmd)
    return cmd
  end

  function emitter:flush()
    local out = M.coalesceCommands(self.pendingCommands)
    self.pendingCommands = {}
    return out
  end

  return emitter
end

function M.newTreeState()
  return {
    children = {},
    nodesById = {},
  }
end

local function ensureTree(tree)
  if type(tree) ~= "table" then
    tree = nil
  end
  if tree == nil then
    tree = M.newTreeState()
  end
  tree.children = tree.children or {}
  tree.nodesById = tree.nodesById or {}
  return tree
end

local function ensureNode(tree, id)
  local node = tree.nodesById[id]
  if node == nil then
    node = {
      _id = id,
      children = {},
    }
    tree.nodesById[id] = node
  elseif node.children == nil then
    node.children = {}
  end
  return node
end

local function applyUpdate(node, cmd)
  node.props = node.props or {}

  for _, key in ipairs(cmd.removeKeys or {}) do
    node.props[key] = nil
  end

  if node.props.style == nil and ((cmd.props or {}).style ~= nil or #(cmd.removeStyleKeys or {}) > 0) then
    node.props.style = {}
  end

  if node.props.style ~= nil then
    for _, key in ipairs(cmd.removeStyleKeys or {}) do
      node.props.style[key] = nil
    end
  end

  for key, value in pairs(cmd.props or {}) do
    if key == "style" and type(value) == "table" then
      local merged = deep_clone(node.props.style or {})
      for style_key, style_value in pairs(value) do
        merged[style_key] = deep_clone(style_value)
      end
      node.props.style = merged
    else
      node.props[key] = deep_clone(value)
    end
  end

  if cmd.hasHandlers ~= nil then
    node.hasHandlers = cmd.hasHandlers
  end
  if cmd.handlerNames ~= nil then
    node.handlerNames = deep_clone(cmd.handlerNames)
  end
  if cmd.handlerMeta ~= nil then
    node.handlerMeta = deep_clone(cmd.handlerMeta)
  end
end

local function removeFromParentList(list, node)
  remove_value(list, node)
end

function M.applyCommand(tree, cmd)
  tree = ensureTree(tree)
  if type(cmd) ~= "table" then
    return tree
  end

  if cmd.op == "CREATE_TEXT" then
    local node = ensureNode(tree, cmd.id)
    node.type = nil
    node.props = node.props or {}
    node.children = nil
    node._text = tostring(cmd.text)
    return tree
  end

  if cmd.op == "CREATE" then
    local node = ensureNode(tree, cmd.id)
    node.type = cmd.type
    node.props = deep_clone(cmd.props or {})
    node.children = node.children or {}
    node._text = nil
    if cmd.hasHandlers ~= nil then
      node.hasHandlers = cmd.hasHandlers
    end
    if cmd.handlerNames ~= nil then
      node.handlerNames = deep_clone(cmd.handlerNames)
    end
    if cmd.handlerMeta ~= nil then
      node.handlerMeta = deep_clone(cmd.handlerMeta)
    end
    return tree
  end

  if cmd.op == "APPEND" then
    local parent = ensureNode(tree, cmd.parentId)
    local child = ensureNode(tree, cmd.childId)
    parent.children = parent.children or {}
    insert_before(parent.children, child)
    child._parentId = parent._id
    return tree
  end

  if cmd.op == "APPEND_TO_ROOT" then
    local child = ensureNode(tree, cmd.childId)
    insert_before(tree.children, child)
    child._parentId = nil
    return tree
  end

  if cmd.op == "INSERT_BEFORE" then
    local parent = ensureNode(tree, cmd.parentId)
    local child = ensureNode(tree, cmd.childId)
    local before = tree.nodesById[cmd.beforeId]
    parent.children = parent.children or {}
    insert_before(parent.children, child, before)
    child._parentId = parent._id
    return tree
  end

  if cmd.op == "INSERT_BEFORE_ROOT" then
    local child = ensureNode(tree, cmd.childId)
    local before = tree.nodesById[cmd.beforeId]
    insert_before(tree.children, child, before)
    child._parentId = nil
    return tree
  end

  if cmd.op == "UPDATE" then
    local node = ensureNode(tree, cmd.id)
    applyUpdate(node, cmd)
    return tree
  end

  if cmd.op == "UPDATE_TEXT" then
    local node = ensureNode(tree, cmd.id)
    node._text = tostring(cmd.text)
    if type(node.children) == "table" and node.children[1] and node.children[1]._text ~= nil then
      node.children[1]._text = tostring(cmd.text)
    end
    return tree
  end

  if cmd.op == "REMOVE" then
    local parent = tree.nodesById[cmd.parentId]
    local child = tree.nodesById[cmd.childId]
    if parent and parent.children and child then
      removeFromParentList(parent.children, child)
    end
    return tree
  end

  if cmd.op == "REMOVE_FROM_ROOT" then
    local child = tree.nodesById[cmd.childId]
    if child then
      removeFromParentList(tree.children, child)
    end
    return tree
  end

  return tree
end

function M.applyCommands(tree, commands)
  tree = ensureTree(tree)
  for _, cmd in ipairs(commands or {}) do
    M.applyCommand(tree, cmd)
  end
  return tree
end

return M
